src/backchannel_logout.rs

```diff
@@ -0,0 +1,469 @@
+//! This module implements components needed for [back-channel logout]
+//!
+//! [back-channel logout]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+
+use std::collections::HashMap;
+
+use chrono::{DateTime, Utc};
+use serde::{de::Error, Deserialize};
+
+use crate::{
+    types::helpers::{deserialize_string_or_vec, serde_utc_seconds},
+    types::SessionIdentifier,
+    Audience, IssuerUrl, SubjectIdentifier,
+};
+
+/// The Logout Token as defined in [section 2.4] of the [OpenID Connect Back-Channel Logout spec][1]
+///
+/// [section 2.4]: <https://openid.net/specs/openid-connect-backchannel-1_0.html#LogoutToken>
+/// [1]: <https://openid.net/specs/openid-connect-backchannel-1_0.html>
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LogoutToken {
+    /// The issuer of this token
+    iss: IssuerUrl,
+    /// The audience this token is intended for
+    aud: Vec<Audience>,
+    /// Time at which this token was issued
+    iat: DateTime<Utc>,
+    /// The unique identifier for this token. This can be used to detect
+    /// replay attacks.
+    jti: String,
+    identifier: Identifier,
+    events: HashMap<String, serde_json::Value>,
+}
+
+impl LogoutToken {
+    /// The `iss` claim
+    pub fn issuer(&self) -> &IssuerUrl {
+        &self.iss
+    }
+
+    /// The `aud` claim
+    pub fn audiences(&self) -> impl Iterator<Item = &Audience> {
+        self.aud.iter()
+    }
+
+    /// The `iat` claim
+    pub fn issue_time(&self) -> DateTime<Utc> {
+        self.iat
+    }
+
+    /// The `jti` claim. It's the unique identifier for this token and can be
+    /// used to detect replay attacks.
+    pub fn jti(&self) -> &str {
+        &self.jti
+    }
+
+    /// As per spec, a [`LogoutToken`] MUST either have the `sub`  or `sid`
+    /// claim and MAY contain both. You can match the [`Identifier`] to detect
+    /// which claims are present.
+    pub fn identifier(&self) -> &Identifier {
+        &self.identifier
+    }
+
+    /// A [`LogoutToken`] is compatible with the [SET standard from RFC 8417][1]
+    ///
+    /// [1]: <https://www.rfc-editor.org/info/rfc8417>
+    pub fn events(&self) -> &HashMap<String, serde_json::Value> {
+        &self.events
+    }
+}
+impl<'de> Deserialize<'de> for LogoutToken {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value: serde_json::Value = serde_json::Value::deserialize(deserializer)?;
```

Comment:
If I'm reading the spec correctly, I think this struct should be parsed as a JWT rather than raw JSON, using similar code to `IdToken`:
> OPs send a JWT similar to an ID Token to RPs called a Logout Token to request that they log out.

The code should also verify the JWT signature:
> A Logout Token MUST be signed and MAY also be encrypted

I'd suggest structuring the code similarly to what this crate does for ID tokens:
- `LogoutTokenClaims` struct similar to `IdTokenClaims` (implementing both `Deserialize` and `Serialize` so that both the RP and OP code paths are supported)
- `LogoutToken` struct similar to `IdToken` that wraps `JsonWebToken`
- `LogoutTokenVerifier` struct similar to `IdTokenVerifier` or `UserInfoVerifier` that contains a `JwtClaimsVerifier` and makes sure that a `nonce` is omitted. I think I would just add an optional `nonce` field (annotated with `serde(skip_serializing)` so that it doesn't get serialized by an OP) to `LogoutTokenClaims` have the verifier check that it's `None`. There are some other required verification steps in the [spec](https://openid.net/specs/openid-connect-backchannel-1_0.html#Validation), including for `iat` (see `IdTokenVerifier::iat_verifier_fn`). `JwtClaimsVerifier` will take care of `iss` and `aud`.

Reply:
> If I'm reading the spec correctly, I think this struct should be parsed as a JWT rather than raw JSON

You're totally right and it's because I planned on implementing this a bit differently than `IdToken` but guess I'll better keep the style of this crate then.

Reply:
I'm not feeling so good implementing an extra `LogoutTokenVerifier` because it's basically copy-paste except for the `nonce` part. Maybe there's a way to be more generics about this?

Reply:
hmm... I tried to move all of the shared JWT verification logic to `JwtClaimsVerifier`. The duplicated part here would mostly be the factory methods (e.g., `new_public_client`) and getters/setters I think. I'm not sure how to get around that extra boilerplate. There shouldn't be too much duplicated _logic_ though, other than verifying `iat`, which is simple enough.

I think I'd start by implementing this verifier with all of the needed functionality, and then maybe it'll be clear whether there are any opportunities to factor out common code.

Reply:
Alright then, I'll see what I can do tomorrow.